src/pevm.rs

```diff
@@ -263,10 +266,16 @@ impl Pevm {
                         _ => unreachable!(),
                     }
                     // Assert that evaluated nonce is correct when address is caller.
-                    debug_assert!(
-                        tx.caller != address || tx.nonce.map_or(true, |n| n + 1 == info.nonce)
-                    );
-
+                    if tx.caller == address {
+                        if let Some(nonce) = tx.nonce {
+                            if nonce + 1 != info.nonce {
```

Comment:
For consistency can do `if info.nonce - 1 != expected_nonce ` instead of `+ 1` here and `- 1` below. Alternatively can invest in an intermediate `let observed_nonce = info.nonce - 1`.

Reply:
I will change to `info.nonce.saturating_sub(1)`. Although it's likely that `info.nonce > 0`, I cannot prove that. Note that the expression `0u64 - 1` will cause a panic.

Reply:
Release builds will wrap instead of panicking, but yeah, it's best practice to use `saturating_x` for numerical operations like this.

Reply:
changed to `.saturating_sub(1)`

Reply:
Just had another look and `saturating_x` is actually not the solution? It still allows `tx_nonce == 0` and `nonce == 0` to pass the check. The best solution to not assume the correctness of other components is to check if `nonce` is zero first and return `UnreachableError` (or something better) if it is. Then `-` would satisfy our check.

Reply:
> Just had another look and `saturating_x` is actually not the solution? It still allows `tx_nonce == 0` and `nonce == 0` to pass the check. The best solution to not assume the correctness of other components is to check if `nonce` is zero first and return `UnreachableError` (or something better) if it is. Then `-` would satisfy our check.

Is this ok?

```
// Assert that evaluated nonce is correct when address is caller.
if tx.caller == address {
    if let Some(tx_nonce) = tx.nonce {
        if info.nonce == 0 {
            return Err(PevmError::UnreachableError);
        }
        if tx_nonce != info.nonce - 1 {
            // TODO: Consider falling back to sequential instead
            return Err(PevmError::NonceMismatch {
                tx_idx: *tx_idx,
                tx_nonce,
                executed_nonce: info.nonce - 1,
            });
        }
    }
}
```

------------------

*Another idea*: Can I use `.wrapping_sub(1)`? It's not hard to decipher `NonceMismatch { tx_nonce: 0, executed_nonce: 18446744073709551615 }` though.

Reply:
Maybe this would add more context to the check:
```rust
let executed_nonce = if nonce == 0 {
    return Err(PevmError::UnreachableError);
} else {
    nonce - 1
};
```

Or we can move the `UnreachableError` check to right after the `match` on `MemoryEntry`, and only proceed to `if tx.caller == address {` if that check passes.